modules/index/src/main/java/org/apache/ignite/internal/index/IndexCollector.java

```diff
@@ -0,0 +1,296 @@
+            = new ConcurrentSkipListMap<>();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    /** Constructor. */
+    IndexCollector(CatalogService catalogService) {
+        this.catalogService = catalogService;
+
+        addListeners();
+    }
+
+    /** Recovers internal structures on node recovery. */
+    void recover() {
+        inBusyLock(busyLock, () -> {
+            // It is expected that the methods will be called only on recovery, when the deploy of metastore watches has not yet occurred.
+            int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+            int latestCatalogVersion = catalogService.latestCatalogVersion();
+
+            // At the moment, we will only use tables from the latest version, since so far only replicas for them are started on the node.
+            for (CatalogTableDescriptor table : catalogService.tables(latestCatalogVersion)) {
+                for (int catalogVersion = earliestCatalogVersion; catalogVersion < latestCatalogVersion; catalogVersion++) {
+                    int tableId = table.id();
+                    int nextCatalogVersion = catalogVersion + 1;
+
+                    List<CatalogIndexDescriptor> tableIndexes = catalogService.indexes(catalogVersion, tableId);
+
+                    if (tableIndexes.isEmpty()) {
+                        // Table does not exist yet.
+                        continue;
+                    }
+
+                    List<CatalogIndexDescriptor> nextCatalogVersionTableIndexes = catalogService.indexes(nextCatalogVersion, tableId);
+
+                    assert !nextCatalogVersionTableIndexes.isEmpty()
+                            : String.format("Table should not be dropped: [catalogVersion=%s, tableId=%s]", nextCatalogVersion, tableId);
+
+                    for (CatalogIndexDescriptor tableIndex : tableIndexes) {
+                        if (tableIndex.available() && !contains(nextCatalogVersionTableIndexes, tableIndex)) {
+                            addDroppedAvailableIndex(tableIndex, nextCatalogVersion);
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        droppedAvailableTableIndexes.clear();
+    }
+
+    /**
+     * Collects a list of table indexes that will need to be used for an update operation in RW a transaction. List consists of all indexes
+     * (available and registered) on the requested catalog version, as well as all dropped available indexes from previous catalog versions.
+     *
+     * <p>Returned list is sorted by {@link CatalogObjectDescriptor#id()}. It is expected that the table exists at the time the method is
```

Comment:
What does 'at the time' mean? Is it about the table existing in the specified catalog version?

Reply:
Fix it